avail-subxt/examples/check_finality.rs

```diff
@@ -0,0 +1,267 @@
+pub struct SignedPrecommit {
+	pub precommit: Precommit,
+	/// The signature on the message.
+	pub signature: Signature,
+	/// The Id of the signer.
+	pub id: EdPublic,
+}
+#[derive(Clone, Debug, Decode, Deserialize)]
+pub struct Commit {
+	pub target_hash: H256,
+	/// The target block's number.
+	pub target_number: u32,
+	/// Precommits for target block or any block after it that justify this commit.
+	pub precommits: Vec<SignedPrecommit>,
+}
+
+#[derive(Clone, Debug, Decode)]
+pub struct GrandpaJustification {
+	pub round: u64,
+	pub commit: Commit,
+	pub votes_ancestries: Vec<Header>,
+}
+
+impl<'de> Deserialize<'de> for GrandpaJustification {
+	fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+	where
+		D: serde::Deserializer<'de>,
+	{
+		let encoded = bytes::deserialize(deserializer)?;
+		Self::decode(&mut &encoded[..])
+			.map_err(|codec_err| D::Error::custom(format!("Invalid decoding: {:?}", codec_err)))
+	}
+}
+
+#[derive(Clone, Debug, Decode)]
+pub enum Messages {
+	Justification(GrandpaJustification),
+	ValidatorSetChange((Vec<EdPublic>, u64)),
+	NewHeader(Header),
+}
+
+#[tokio::main]
+async fn main() {
+	let url = "ws://localhost:9944";
+	let subxt_client = build_client(url).await.unwrap();
+
+	// Subscribe to finalized headers (actually, finalized chain's head headers, some final headers may be skipped).
+	let mut header_subscription = subxt_client
+		.rpc()
+		.subscribe_finalized_block_headers()
+		.await
+		.unwrap();
+
+	// Current set of authorities, implicitly trusted, fetched from grandpa runtime.
+	let grandpa_valset_raw = subxt_client
+		.runtime_api()
+		.at(None)
+		.await
+		.unwrap()
+		.call_raw("GrandpaApi_grandpa_authorities", None)
+		.await
+		.unwrap();
+
+	// Decode result to proper type - ed25519 public key and u64 weight.
+	let grandpa_valset: Vec<(EdPublic, u64)> =
+		Decode::decode(&mut &grandpa_valset_raw[..]).unwrap();
+
+	// Drop weights, as they are not currently used.
+	let mut validator_set: Vec<EdPublic> = grandpa_valset.iter().map(|e| e.0).collect();
+
+	// Set ID is acquired in a separate storage query. It is necessary, because it is a part of message being signed.
+
+	// Form a query key for storage
+	let set_id_key = api::storage().grandpa().current_set_id();
+	// Fetch the set ID from storage at current height
+	let mut set_id = subxt_client
+		.storage()
+		// None means current height
+		.at(None)
+		.await
+		.unwrap()
+		.fetch(&set_id_key)
+		.await
+		.unwrap()
+		.unwrap();
+
+	println!("Current set: {:?}", validator_set);
+
+	// Forming a channel for sending any relevant events gathered asynchronously through Substrate WS API.
+	let (msg_sender, mut msg_receiver) = unbounded_channel::<Messages>();
+
+	// Task that produces headers and new validator sets
+	tokio::spawn({
+		let subxt_client = subxt_client.clone();
+		let msg_sender = msg_sender.clone();
+		async move {
+			while let Some(Ok(header)) = header_subscription.next().await {
```

Comment:
Does this also suffer the same problem of jumping blocks in case of delayed finality? The `subxt` examples also do it like this, so I'm guessing it gets handled internally?

Reply:
Although, on second thought, this particular case is an exception. Auth set change containing blocks are always finalized. But still my original question remains for my general understanding.

Reply:
Yes, this also suffers the same issue of skipping blocks. But for verifying finality and changes in auth set, it's not a problem, since changing authset is always in a block signed by supermajorty, so it's always properly verified. And I did test it too. I paused finality until the change of session/new authset and resumed it after and I received a header containing changes in auth set and resumed normally.

In this example I don't backtrack the headers till the last checkpoint, because there's no need to. If the supermajority agrees the block is the new head, that's enough. For sampling and appdata purposes, this will have to be done, of course.

Reply:
Got it. Makes sense. Thanks!